### PR TITLE
Heater target reached notifications

### PR DESCRIPTION
**Pull request by encetamasb**

### CODE REVIEW

File: web/app/heater_trackers.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+                event = HeaterEvent(type=etype, state=heater)
+
+        ret.append((tracker, dirty, event))
+
+    return ret
+
+
+def update_heater_trackers(printer: Printer, temp_data: Dict) -> None:
+    heaters = list(parse_states(temp_data).values())
+    trackers = list(printer.heatertracker_set.all())
```
[the remark this answers is not in the document]
Comment (encetamasb) [reply web/app/heater_trackers.py:134]: @kennethjiang (1) redis: ok for me, btw dont you have AOF enabled in redis? migration can be done without downtime as well. (2) or a json field? can be put into Printer or into a model with a OneToOneField

[the remark this answers is not in the document]
Comment (kennethjiang) [reply web/app/heater_trackers.py:134]: I know redis can be migrated without downtime. But every time I did it I was still nervous lol.

If we put it as a json field in Printer, we will still need to refresh_from_db, as the `printer` object is retrieved from the db when the WS connection was first created, whereas messages will come many times after that. Let me know if you have a good db solution without significant db load increase.

Also my apologies for not sending my comments earlier. I did the comments yesterday but forgot to "submit the review". Have made the same mistake quite a few times now... :(

[the remark this answers is not in the document]
Comment (encetamasb) [reply web/app/heater_trackers.py:134]: @kennethjiang Its not super necessary to refresh from db everytime. We can keep a cache in Printer object (printer._heater_trackers  = ...) and save/delete trackers only when item is dirty/obsolete. This needs only small changes (fetch from printer or db function, some extra error handling, updating cache in printer object).

[the remark this answers is not in the document]
Comment (kennethjiang) [reply web/app/heater_trackers.py:134]: This is actually a good point, by taking advantage of the fact that one WS connection will be always on the same server.

Any scenario that this cache will be accessed/modified on potentially multiple servers now or in the future?

[the remark this answers is not in the document]
Comment (kennethjiang) [reply web/app/heater_trackers.py:134]: Realized one scenario: what if the plugin gets disconnected somehow (loss of connection, or reboot) and reconnects to a different server? We are currently running 12 websocket servers (containers) in prod so the odds to connect to the same server is quite low unless we start to do some session affinity, which will add a whole layer of complexity to our operations...

[the remark this answers is not in the document]
Comment (encetamasb) [reply web/app/heater_trackers.py:134]: @kennethjiang i think there cannot be any problem as far as there is only one connection per printer. When plugin disconnects, connection is closed and handler should no longer update tracker states.

 What if tracker states are updated silently out of loop? Thats a problem but can be solved with a magic value (updates are succesful only when magic value matches, .. Model.objects.filter(magic=v).update(..) > 0 .. when update does not touch any record then states must be refetched first and retry).

[the remark this answers is not in the document]
Comment (kennethjiang) [reply web/app/heater_trackers.py:134]: It sounds a bit hacky (the reliance on the requirement that only 1 connection from the plugin can be made, and the magic number). But it sounds like the reasonable trade-off given the amount of db/redis resource it'll save by not refreshing all the time.

[the remark this answers is not in the document]
Comment (kennethjiang) [reply web/app/heater_trackers.py:134]: Also what are the tradeoffs between onetoone relationship and json field? I have not used json field myself. One thing to throw in here is we are using sqlite light for the users who run their own server. So we want to make sure json field works in sqlite the same way as postgres if we go that route.

[the remark this answers is not in the document]
Comment (encetamasb) [reply web/app/heater_trackers.py:134]: There is no problem with JSONField in sqlite (at least with that library what's been already imported into app.models). Its mapped as a text column. Tradeoffs.. no interesting ones imo, I think this is mostly about "style".

[the remark this answers is not in the document]
Comment (encetamasb) [reply web/app/heater_trackers.py:134]: The magic/version system is a good idea though in my opinion, even if its not vital in this case probably (can anything go really wrong? Worst case is 2x notification in same cases). And its not that hacky :)

[the remark this answers is not in the document]
Comment (kennethjiang) [reply web/app/heater_trackers.py:134]: Good point about magic number is not a hack.
Thank you for the tradeoff details. Feel free to make your own call.

[the remark this answers is not in the document]
Comment (encetamasb) [reply web/app/heater_trackers.py:134]: ok, will update repo today

**Pull request merged**